# Practice 5

### Import libraries

In [81]:
#Import libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
#Cross validation 
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
#Confusion matrix
from sklearn.metrics import (confusion_matrix, accuracy_score)
# to import the cancer patients dataset
from sklearn.datasets import load_breast_cancer
#Suport vector machine
from sklearn import svm
from sklearn.svm import SVC
#Decicion trees
from sklearn import tree
#Naive Bayes
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
#Nearest Neighbors
from sklearn.neighbors import KNeighborsClassifier
#PCA
# to standardize the features
from sklearn.preprocessing import StandardScaler 
# to apply PCA
from sklearn.decomposition import PCA
#Gradient Boosting Classifier
from sklearn.ensemble import GradientBoostingClassifier
#Bagging Classifier
from sklearn.ensemble import BaggingClassifier
#Libraries for stacking
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import StackingClassifier
#Random Forest
from sklearn.ensemble import RandomForestClassifier
#Ada Boost
from sklearn.ensemble import AdaBoostClassifier

## Data processing

In [3]:
variables, target = load_breast_cancer(return_X_y=True,as_frame=True) #load dataset

In [4]:
#features = dataframe of features variables
#target= target variable (categorical)
features=variables.iloc[0:,[0,1,2,3,4,5,6,7,8,9]]
target1=pd.DataFrame(target)
target1.rename({'target': 'Diagnosis'}, axis=1, inplace=True)

In [4]:
features

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  
0                   0.07871  
1                   0.05667  
2                   0.05999  
3                   0.09744  
4                   0.05883  
..                      ...  
564                 0.05623  
565                 0.05533  
566                 0.05648  
567                 0.07016  
568                 0.05884  

[569 rows x 10 columns]

In [5]:
target1

Diagnosis
0            0
1            0
2            0
3            0
4            0
..         ...
564          0
565          0
566          0
567          0
568          1

[569 rows x 1 columns]

## Cross validation

In [6]:
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size=0.8, random_state=1)

## Aplying meta-learn 

### Boosting: 
This algorithm builds an additive model in a forward stage-wise fashion; it allows for the optimization of arbitrary differentiable loss functions. In each stage n_classes_ regression trees are fit on the negative gradient of the loss function, e.g. binary or multiclass log loss. Binary classification is a special case where only a single regression tree is induced.

In [22]:
#Boosting with default values
GBC = GradientBoostingClassifier(n_estimators=100, learning_rate=0.2,max_depth=1, random_state=0)
y_prediction = cross_val_predict(GBC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 1 0 1 0
 1 1]
Confusion matrix: [[42  3]
 [ 2 66]]
True negative: 42
False negative: 2
True positive: 66
False positive: 3
Accuracy: 0.9557522123893806
Sensitivity: 0.9705882352941176
Specificity: 4.0


##### Trying to better the model.

In [24]:
#Boosting with exponential loss and he function to measure the quality of a split is sqared error
GBC = GradientBoostingClassifier(loss='exponential',n_estimators=200, 
                                 learning_rate=1.0,max_depth=1, criterion='squared_error',random_state=0)
y_prediction = cross_val_predict(GBC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 1 1 0 0 1 1 0 0 1 0 1 1 1 1 1 0 1 0
 1 1]
Confusion matrix: [[41  4]
 [ 1 67]]
True negative: 41
False negative: 1
True positive: 67
False positive: 4
Accuracy: 0.9557522123893806
Sensitivity: 0.9852941176470589
Specificity: 5.0


| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
| Boosting (default values)| 0.95 |  0.97 | 4 |
| Boosting (loss=exponential) | 0.95 | 0.98 | 5 |


### Baggin:
Bagging is an ensemble machine learning algorithm that combines the predictions from many decision trees. It is easy to implement with few key hyperparameters and sensible heuristics for configuring these hyperparameters. Bagging performs well in general and provides the basis for a whole field of ensemble of decision tree algorithms such as the popular random forest and extra.

In [29]:
#Baggin clasification with default values
BAGGIN = BaggingClassifier(estimator=SVC(),n_estimators=10, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 1 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 1 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[30 15]
 [ 2 66]]
True negative: 30
False negative: 2
True positive: 66
False positive: 15
Accuracy: 0.8495575221238938
Sensitivity: 0.9705882352941176
Specificity: 16.0


##### Trying to better the model.

In [34]:
#Baggin clasification with Decision Tree estimator, The function to measure the quality of a split is entropy
d_tree = tree.DecisionTreeClassifier(criterion='entropy')
BAGGIN = BaggingClassifier(estimator=d_tree,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 0 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[40  5]
 [ 5 63]]
True negative: 40
False negative: 5
True positive: 63
False positive: 5
Accuracy: 0.911504424778761
Sensitivity: 0.9264705882352942
Specificity: 6.0


In [35]:
#Baggin clasification with Naive Bayes estimator
gnb = GaussianNB()
BAGGIN = BaggingClassifier(estimator=gnb,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 0
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[41  4]
 [ 2 66]]
True negative: 41
False negative: 2
True positive: 66
False positive: 4
Accuracy: 0.9469026548672567
Sensitivity: 0.9705882352941176
Specificity: 5.0


In [38]:
#Baggin clasification with Bernoulli Naive Bayes estimator
BNB= BernoulliNB()
BAGGIN = BaggingClassifier(estimator=BNB,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1]
Confusion matrix: [[ 0 45]
 [ 1 67]]
True negative: 0
False negative: 1
True positive: 67
False positive: 45
Accuracy: 0.5929203539823009
Sensitivity: 0.9852941176470589
Specificity: nan


/tmp/ipykernel_2319/1670565986.py:23: RuntimeWarning: invalid value encountered in long_scalars
  SPE=TN/TN+FP


In [79]:
#Baggin clasification with KNeighbors Classifier estimator
knn = KNeighborsClassifier(n_neighbors=10, algorithm='kd_tree')
BAGGIN = BaggingClassifier(estimator=knn,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 1 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 1 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[36  9]
 [ 3 65]]
True negative: 36
False negative: 3
True positive: 65
False positive: 9
Accuracy: 0.8938053097345132
Sensitivity: 0.9558823529411765
Specificity: 10.0


In [43]:
#Baggin clasification with support vector machine estimator, linear kernel
SVM = svm.SVC(kernel='linear', C=1)
BAGGIN = BaggingClassifier(estimator=SVM,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 1 0 0 1 1 0 1 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 0 0 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[40  5]
 [ 5 63]]
True negative: 40
False negative: 5
True positive: 63
False positive: 5
Accuracy: 0.911504424778761
Sensitivity: 0.9264705882352942
Specificity: 6.0


In [44]:
#Baggin clasification with support vector machine estimator polynomial kernel
SVM = svm.SVC(kernel='poly', C=1)
BAGGIN = BaggingClassifier(estimator=SVM,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0 1 0 0 1 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 1 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 1 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[31 14]
 [ 2 66]]
True negative: 31
False negative: 2
True positive: 66
False positive: 14
Accuracy: 0.8584070796460177
Sensitivity: 0.9705882352941176
Specificity: 15.0


In [46]:
#Baggin clasification with support vector machine estimator, SVC LINEAR kernel
SVM =svm.LinearSVC(C=1, max_iter=10000, dual="auto")
BAGGIN = BaggingClassifier(estimator=SVM,n_estimators=100, random_state=0)
y_prediction = cross_val_predict(BAGGIN, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[40  5]
 [ 2 66]]
True negative: 40
False negative: 2
True positive: 66
False positive: 5
Accuracy: 0.9380530973451328
Sensitivity: 0.9705882352941176
Specificity: 6.0


| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
| Baggin (default values) | 0.84 |  0.97 | 16 |
| Baggin (Decision Tree, loss=entropy) | 0.91 | 0.92 | 6 |
| Baggin (Gaussian Naive Bayes) | 0.94 |  0.97 | 5 |
| Baggin (Bernoulli Naive Bayes | 0.59 |  0.98 | 5 |
| Baggin (K Neighbors) | 0.89 |  0.95 | 10 |
| Baggin (SVM linear kernel) | 0.91 |  0.92 | 6 |
| Baggin (linear SVC) | 0.93 |  0.97 | 6 |
| Baggin (SVM polynomial kernel) | 0.85 |  0.97 | 6 |


### Stacking:
Stack of estimators with a final classifier. Stacked generalization consists in stacking the output of individual estimator and use a classifier to compute the final prediction. Stacking allows to use the strength of each individual estimator by using their output as input of a final estimator. Note that estimators_ are fitted on the full X while final_estimator_ is trained using cross-validated predictions of the base estimators using cross_val_predict.

In [56]:
#Stacking clasification with Random Forest, linear SVC and logistic regresion. 
estimators=[
     ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
     ('svr', make_pipeline(StandardScaler(),
     LinearSVC(dual="auto", random_state=42)))]
#Creating the model
STC=StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())
STC.fit(x_train, y_train).score(x_test, y_test)

0.9144736842105263

##### Trying to better the model.

In [60]:
#Stacking clasification with Random Forest, poly SVM and logistic regresion. 

#Weak learners
estimators=[
     ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
     ('svm',svm.SVC(kernel='poly', C=1) ,
     )]
#Creating the model
STC=StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())

y_prediction = cross_val_predict(STC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 1 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 1 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[37  8]
 [ 3 65]]
True negative: 37
False negative: 3
True positive: 65
False positive: 8
Accuracy: 0.9026548672566371
Sensitivity: 0.9558823529411765
Specificity: 9.0


In [62]:
#Stacking clasification with Random Forest, poly SVM, Gaussian Naive Bayes, and logistic regresion. 

#Weak learners
estimators=[
     ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
     ('svm',svm.SVC(kernel='poly', C=1) ,
     GaussianNB())]
#Creating the model
STC=StackingClassifier(estimators=estimators, final_estimator=tree.DecisionTreeClassifier(criterion='entropy'))

y_prediction = cross_val_predict(STC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 0 1 1 1 1 0 0 1 1 1 1 1 0 0 0 1 1 0 0 0 1 1 0 0 1 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 0 1 0 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[35 10]
 [ 7 61]]
True negative: 35
False negative: 7
True positive: 61
False positive: 10
Accuracy: 0.8495575221238938
Sensitivity: 0.8970588235294118
Specificity: 11.0


In [72]:
#Stacking clasification with Random Forest, poly SVM, Gaussian Naive Bayes.

#Weak learners
estimators=[
     ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
     ('svm',svm.SVC(kernel='rbf', C=1) ,
     GaussianNB())]
#Creating the model
STC=StackingClassifier(estimators=estimators, final_estimator=GaussianNB())

y_prediction = cross_val_predict(STC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 0 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[39  6]
 [ 4 64]]
True negative: 39
False negative: 4
True positive: 64
False positive: 6
Accuracy: 0.911504424778761
Sensitivity: 0.9411764705882353
Specificity: 7.0


| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
| Stacking (Random Forest, linear SVM, logistic regresion) | 0.90 |  0.95 | 9 |
| Stacking (Random Forest, poly SVM, Gaussian Naive Bayes, logistic regresion) | 0.84 | 0.89 | 11 |
| Stacking (Random Forest, poly SVM, Gaussian Naive Bayes, tree decision) | 0.91 |  0.94 | 7 |


### RandomForestClassifier
A random forest classifier, is a meta estimator that fits a number of decision tree classifiers on various sub-samples of the dataset and uses averaging to improve the predictive accuracy and control over-fitting. The sub-sample size is controlled with the max_samples parameter if bootstrap=True (default), otherwise the whole dataset is used to build each tree

In [75]:
#Random Forest with default values
# Creating the model
RFC = RandomForestClassifier(random_state=0)

y_prediction = cross_val_predict(RFC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 0 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[39  6]
 [ 5 63]]
True negative: 39
False negative: 5
True positive: 63
False positive: 6
Accuracy: 0.9026548672566371
Sensitivity: 0.9264705882352942
Specificity: 7.0


##### Trying to better the model.

In [76]:
#Random Forest with other parameters
# Creating the model
RFC = RandomForestClassifier(criterion='entropy',max_depth=4,random_state=0,max_features='log2')

y_prediction = cross_val_predict(RFC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 1 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[39  6]
 [ 4 64]]
True negative: 39
False negative: 4
True positive: 64
False positive: 6
Accuracy: 0.911504424778761
Sensitivity: 0.9411764705882353
Specificity: 7.0


In [78]:
#Random Forest with other parameters
# Creating the model
RFC = RandomForestClassifier(criterion='log_loss',max_depth=10,random_state=2,max_features=None)

y_prediction = cross_val_predict(RFC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 0 0 1 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[39  6]
 [ 3 65]]
True negative: 39
False negative: 3
True positive: 65
False positive: 6
Accuracy: 0.9203539823008849
Sensitivity: 0.9558823529411765
Specificity: 7.0


| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
| Random Forest (default values) | 0.90 |  0.92 | 7 |
| Random Forest (criterion=entropy) | 0.91 | 0.94 | 7 |
| Random Forest (criterion=log loss) | 0.92 |  0.95 | 7 |

###  AdaBoost:

An AdaBoost classifier is a meta-estimator that begins by fitting a classifier on the original dataset and then fits additional copies of the classifier on the same dataset but where the weights of incorrectly classified instances are adjusted such that subsequent classifiers focus more on difficult cases.

In [82]:
# AdaBoost classifier with default values
# Creating the model
ABC=AdaBoostClassifier(n_estimators=100, random_state=0)

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 0 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 1 0 1 0
 1 1]
Confusion matrix: [[41  4]
 [ 1 67]]
True negative: 41
False negative: 1
True positive: 67
False positive: 4
Accuracy: 0.9557522123893806
Sensitivity: 0.9852941176470589
Specificity: 5.0


##### trying to improve the model

In [95]:
# AdaBoost classifier with Linear SVC estimator
# Creating the model
SVM =svm.LinearSVC(C=1, max_iter=10000, dual="auto")
ABC=AdaBoostClassifier(estimator=SVM, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 1 1 1 1 1 0 0 1 1 1 0 1 0 0 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 1
 1 1 0 0 1 1 0 0 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 0 1 0 1 1 0 0 1 0 0 1 1 0
 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 0 1]
Confusion matrix: [[39  6]
 [ 6 62]]
True negative: 39
False negative: 6
True positive: 62
False positive: 6
Accuracy: 0.8938053097345132
Sensitivity: 0.9117647058823529
Specificity: 7.0


In [96]:
# AdaBoost classifier with decision tree estimator
# Creating the model
d_tree = tree.DecisionTreeClassifier(criterion='entropy')
ABC=AdaBoostClassifier(estimator=d_tree, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 0 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 0 1 0 0 0 1 0 1 0 0 1 0
 1 1 0 0 1 1 1 0 1 0 0 1 0 1 1 0 1 1 0 0 1 0 1 1 1 0 0 1 1 0 0 0 0 1 1 1 0
 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 1 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 1 1]
Confusion matrix: [[40  5]
 [ 8 60]]
True negative: 40
False negative: 8
True positive: 60
False positive: 5
Accuracy: 0.8849557522123894
Sensitivity: 0.8823529411764706
Specificity: 6.0


In [97]:
# AdaBoost classifier with gaussian naive bayes
# Creating the model
gnb = GaussianNB()
ABC=AdaBoostClassifier(estimator=gnb, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [0 1 1 0 1 0 1 1 1 1 0 0 1 1 1 0 1 0 1 0 1 1 1 0 1 1 1 0 0 0 1 0 1 0 0 1 0
 1 1 0 0 1 1 1 1 1 0 0 1 0 1 1 1 1 1 0 0 1 1 1 1 1 0 0 1 1 0 0 0 0 0 1 1 0
 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 1 1 1 1 0 0 1 0
 0 0]
Confusion matrix: [[42  3]
 [ 5 63]]
True negative: 42
False negative: 5
True positive: 63
False positive: 3
Accuracy: 0.9292035398230089
Sensitivity: 0.9264705882352942
Specificity: 4.0


In [94]:
# AdaBoost classifier with Bernoulli Naive Bayes
# Creating the model
BNB= BernoulliNB()
ABC=AdaBoostClassifier(estimator=BNB, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(BNB, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1]
Confusion matrix: [[ 0 45]
 [ 0 68]]
True negative: 0
False negative: 0
True positive: 68
False positive: 45
Accuracy: 0.6017699115044248
Sensitivity: 1.0
Specificity: nan


/tmp/ipykernel_2319/2923798029.py:25: RuntimeWarning: invalid value encountered in long_scalars
  SPE=TN/TN+FP


In [98]:
# AdaBoost classifier with SVM kernel='poly'
# Creating the model
SVM = svm.SVC(kernel='poly', C=1)
ABC=AdaBoostClassifier(estimator=SVM, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 0 0 1 1 1 0 0 1 1
 1 1 0 1 1 1 1 1 1 0 1 1 0 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 0 0 1 0 1 1 1 1 1 0 1 1
 1 1]
Confusion matrix: [[21 24]
 [ 0 68]]
True negative: 21
False negative: 0
True positive: 68
False positive: 24
Accuracy: 0.7876106194690266
Sensitivity: 1.0
Specificity: 25.0


In [100]:
# AdaBoost classifier with SVM kernel='rbf'
# Creating the model
SVM = svm.SVC(kernel='rbf', C=1)
ABC=AdaBoostClassifier(estimator=SVM, n_estimators=100, random_state=0,algorithm='SAMME')

y_prediction = cross_val_predict(ABC, x_train, y_train, cv=10)
print('Predicted values:',y_prediction)

#Metrics
#Confusion matrix
CM=confusion_matrix(y_train, y_prediction)
#true negative
TN=CM[0,0]
#false negative
FN=CM[1,0]
#true positive
TP=CM[1,1]
#false positive
FP=CM[0,1]
#Accuracy 
accuracy=accuracy_score(y_train, y_prediction)
#Sensitivity
SEN=TP/(TP+FN)
#Specificity
SPE=TN/TN+FP

print('Confusion matrix:',CM)
print('True negative:',TN)
print('False negative:',FN)
print('True positive:',TP)
print('False positive:',FP)

print('Accuracy:',accuracy) 
print('Sensitivity:',SEN)
print('Specificity:',SPE)

Predicted values: [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1]
Confusion matrix: [[ 0 45]
 [ 0 68]]
True negative: 0
False negative: 0
True positive: 68
False positive: 45
Accuracy: 0.6017699115044248
Sensitivity: 1.0
Specificity: nan


/tmp/ipykernel_2319/3080983698.py:25: RuntimeWarning: invalid value encountered in long_scalars
  SPE=TN/TN+FP


| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
| Adaboost (default values) | 0.95 |  0.98 | 5 |
| Adaboost (linear SVC) | 0.91 | 0.92 | 6 |
| Adaboost (Gaussian Naive Bayes) | 0.92 |  0.92 | 4 |
| Adaboost (Bernoulli Naive Bayes | 0.60 |  1 | x |
| Adaboost (decision tree entropy) | 0.88 |  0.88 | 6 |
| Adaboost (linear SVC) | 0.89 |  0.91 | 7 |
| Adaboost (SVM polynomial kernel) | 0.78 |  1 | 25 |
| Adaboost (SVM RBF kernel) | 0.60 | 1 | x |

## The best classifiers 

| Model | Accuracy | Sensitivity | Specificity |
| --- | --- | --- | --- |
|     |     | Meta ensamble |     |
| Adaboost (default values) | 0.95 |  0.98 | 5 |
| Random Forest (criterion=log loss) | 0.92 |  0.95 | 7 |
| Stacking (Random Forest, poly SVM, Gaussian Naive Bayes, tree decision) | 0.91 |  0.94 | 7 |
| Baggin (Gaussian Naive Bayes) | 0.94 |  0.97 | 5 |
| Boosting (loss=exponential) | 0.95 | 0.98 | 5 |
|     |     | Weak learners |     |
| Suport vector machine | 0.91 |  0.92 | 6 |
| Decision Trees | 0.89 | 0.89 | 6 |
| Naive Bayes | 0.94 |  0.97 | 5 |
| K Neighbors Classifier | 0.88 |  0.94 | 10 |
| Naive Bayes (PCA) | 0.94 |  0.95 | 5 |

## Conclusion
For this case, the best classfiers were Adaboost and Boosting, because the metrics show a better classification and prediction for the target variables he was the diagnosis, which can be a cancer malignan (0) or benignin (1), even better than the weak learner Naive Bayes.